File: app/src/main/java/com/commit451/gitlab/widget/ProjectFeedWidgetPrefs.java
package com.commit451.gitlab.widget;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.bluelinelabs.logansquare.LoganSquare;
import com.commit451.gitlab.model.Account;

import java.io.IOException;

/**
 * The prefs for the feed widget
 */
public class ProjectFeedWidgetPrefs {

    public static String FILE_NAME = "LabCoatProjectWidgetPrefs";

    private static final String KEY_ACCOUNT = "_account";
    private static final String KEY_PROJECT_FEED_URL = "_feed_url";

    private static SharedPreferences getSharedPrefs(Context context) {
        return context.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE);
    }

    @Nullable
    public static Account getAccount(Context context, int widgetId) {
        String accountsJson = getSharedPrefs(context).getString(widgetId + KEY_ACCOUNT, null);
        if (!TextUtils.isEmpty(accountsJson)) {
            try {
                return LoganSquare.parse(accountsJson, Account.class);
            } catch (IOException e) {
                //why would this ever happen?!?!?1
                getSharedPrefs(context).edit().remove(widgetId + KEY_ACCOUNT).commit();
            }
        }
        return null;
    }

    public static void setAccount(Context context, int widgetId, Account account) {
        try {
            String json = LoganSquare.serialize(account);
            getSharedPrefs(context)
                    .edit()
                    .putString(widgetId + KEY_ACCOUNT, json)
                    .commit();
        } catch (IOException e) {
            //this wont happen! Right?!?!?!
        }
    }

    @Nullable
    public static String getFeedUrl(Context context, int widgetId) {
        return getSharedPrefs(context).getString(widgetId + KEY_PROJECT_FEED_URL, null);
    }

    public static void setFeedUrl(Context context, int widgetId, String url) {
        getSharedPrefs(context)
                .edit()
                .putString(widgetId + KEY_PROJECT_FEED_URL, url)
                .commit();
    }
}

File: app/src/main/java/com/commit451/gitlab/widget/ProjectFeedWidgetPrefs.kt
package com.commit451.gitlab.widget

import android.content.Context
import android.content.SharedPreferences
import android.text.TextUtils

import com.bluelinelabs.logansquare.LoganSquare
import com.commit451.gitlab.model.Account

import java.io.IOException

/**
 * The prefs for the feed widget
 */
object ProjectFeedWidgetPrefs {

    private val FILE_NAME = "LabCoatProjectWidgetPrefs"

    private val KEY_ACCOUNT = "_account"
    private val KEY_PROJECT_FEED_URL = "_feed_url"

    private fun getSharedPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)
    }

    fun getAccount(context: Context, widgetId: Int): Account? {
        val accountsJson = getSharedPrefs(context).getString(widgetId.toString() + KEY_ACCOUNT, null)
        if (!TextUtils.isEmpty(accountsJson)) {
            try {
                return LoganSquare.parse(accountsJson, Account::class.java)
            } catch (e: IOException) {
                //why would this ever happen?!?!?1
                getSharedPrefs(context).edit().remove(widgetId.toString() + KEY_ACCOUNT).commit()
            }

        }
        return null
    }

    fun setAccount(context: Context, widgetId: Int, account: Account) {
        try {
            val json = LoganSquare.serialize(account)
            getSharedPrefs(context)
                    .edit()
                    .putString(widgetId.toString() + KEY_ACCOUNT, json)
                    .commit()
        } catch (e: IOException) {
            //this wont happen! Right?!?!?!
        }

    }

    fun getFeedUrl(context: Context, widgetId: Int): String? {
        return getSharedPrefs(context).getString(widgetId.toString() + KEY_PROJECT_FEED_URL, null)
    }

    fun setFeedUrl(context: Context, widgetId: Int, url: String) {
        getSharedPrefs(context)
                .edit()
                .putString(widgetId.toString() + KEY_PROJECT_FEED_URL, url)
                .commit()
    }
}
